### PR TITLE
SI-7331 importers now import tree positions

### PR DESCRIPTION
**Pull request by jedesah**
There was also an issue where because of the wrapper code that the toolbox adds in order to support a wider range of string inputs than the vanilla compiler, it was necessary to adjust the positions associated with the trees generated by the compiler in order to provide a coherent interface to client code.

Furthermore, an assert was removed from emptyValDef in favour of a more lenient implementation that avoids needless guards against emptyValDef in client code.

### CODE REVIEW

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 1]
```diff
@@ -281,6 +284,20 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
         val wrappedTree = parser.parse()
         throwIfErrors()
         val PackageDef(_, List(ModuleDef(_, _, Template(_, _, _ :: parsed)))) = wrappedTree
+        def adjustPosition(tree: Tree) {
+          if (tree.pos.isDefined) {
+            val wrappedSource = tree.pos.source
+            val adjustedSource = new BatchSourceFile(wrappedSource.file, wrappedSource.content.drop(startWrapperCode.length).dropRight(endWrapperCode.length))
+            val naiveAdjustedOffset = tree.pos.point - startWrapperCode.length
+            // @jedesah writes: This seems buggy to me that we would receive Offset positions that go beyond the scope of the trees we are dealing with,
```
Comment (xeno-by) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:292]: Did this actually happen or you are just guarding from potential bugs?

Comment (jedesah) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:292]: This actually happened. I used my own project's test suite to thoroughly exercise my patch and uncovered a few bugs with my initial implementation. Now that I think of it, I should have written a test case for the compiler every time I encountered buggy behaviour uncovered by the other project.

I can at least add a test case for the above mentioned problem and any others I can remember.

Comment (xeno-by) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:292]: Yes, please do that. Thanks!

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 2]
```diff
@@ -968,7 +968,7 @@ trait Trees extends api.Trees { self: SymbolTable =>
   object emptyValDef extends ValDef(Modifiers(PRIVATE), nme.WILDCARD, TypeTree(NoType), EmptyTree) {
     override def isEmpty = true
     super.setPos(NoPosition)
-    override def setPos(pos: Position) = { assert(false); this }
+    override def setPos(pos: Position) = this
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Trees.scala:971]: @paulp what do you think of this change?

Comment (paulp) [inline src/reflect/scala/reflect/internal/Trees.scala:971]: I think it is brilliance incarnate.

**Comment by densh** [comment]
There is actually a much simpler way to eliminate the need to adjust positions in toolbox: 
1. don't wrap the code snippet all
2. call `templateStats` on a parser object with raw code snippet and call `accept(EOF)` in the end instead of `parse`

This is what repl and scriptRunner are doing for parsing and that's also the way quasiquotes are going to parse code after my next pull request. (see `codeParser` in `ExprTyper.scala` for example)

Considering the fact that extremely similar use cases arise in 4 different contexts (repl, toolbox, quasiquotes, scriptRunner) we might want to generalize this out in some form.

/сс @xeno-by @paulp @jedesah

**Comment by retronym** [comment]
I think @densh's idea should be pursued. I'll close this PR for now in anticipation of his rework.

**Comment by jedesah** [comment]
@densh Glad to hear there is a better solution! Furthermore extracting the common functionality between all these contexts sounds like the thing to do.

Let me know if there is anything I can do to help. I would really like to see positioned trees coming out of the ToolBox in the upcoming release candidate of 2.10.3 as it would make it much easier for me to publish one of my libraries.

**Comment by adriaanm** [comment]
(Note that the PR will need to be reviewed and merged by the end of the week for it to make it into 2.10.3.)

**Pull request closed without merging**